### ISSUE TITLE
[Troubleshoot]: I encountered an unexpected warning during training, and I'm concerned it might affect the training process.

### ISSUE THREAD

**Issue by tobottyx** [RLinf/RLinf#504]
### Problem description

(EmbodiedFSDPActor pid=93386)   var_returns = torch.var(masked_returns)
(EmbodiedFSDPActor pid=93386) /root/autodl-tmp/icr_Rlinf/rlinf/algorithms/losses.py:199: UserWarning: var(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at /pytorch/aten/src/ATen/native/ReduceOps.cpp:1831.)
(EmbodiedFSDPActor pid=93386)   var_returns = torch.var(masked_returns)
(EmbodiedFSDPActor pid=93386) /root/autodl-tmp/icr_Rlinf/rlinf/algorithms/losses.py:199: UserWarning: var(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at /pytorch/aten/src/ATen/native/ReduceOps.cpp:1831.)
(EmbodiedFSDPActor pid=93386)   var_returns = torch.var(masked_returns)
(EmbodiedFSDPActor pid=93386) /root/autodl-tmp/icr_Rlinf/rlinf/algorithms/losses.py:199: UserWarning: var(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at /pytorch/aten/src/ATen/native/ReduceOps.cpp:1831.)
There are a massive number of warnings causing concern for the critic's training stability. What typically causes these NaN values?

### Configuration YAML file

YAML config file:

You can also paste the full config here.
```
defaults:
  - env/libero_10@env.train
  - env/libero_10@env.eval
  - model/pi0@actor.model
  - training_backend/fsdp@actor.fsdp_config
  - override hydra/job_logging: stdout

hydra:
  run:
    dir: .
  output_subdir: null
  searchpath:
    - file://${oc.env:EMBODIED_PATH}/config/

cluster:
  num_nodes: 1
  component_placement:
    actor: 0,
    env: 0,
    rollout: 1

runner:
  task_type: embodied
  logger:
    log_path: "../results"
    project_name: rlinf
    experiment_name: "test_openpi"
    logger_backends: ["tensorboard"] # wandb, swanlab

  max_epochs: 1000
  max_steps: -1

  only_eval: False
  eval_policy_path: null # Optional: .pt file or None, if None, will use the checkpoint in rollout.model.model_path
  val_check_interval: -1
  save_interval: 40
  seq_length: 2048
  max_prompt_length: 50

  resume_dir: null

algorithm:
  normalize_advantages: True
  kl_penalty: kl  # how to estimate kl divergence: kl or kl_penalty
  group_size: 1
  reward_coef: 1.0
  rollout_epoch: 4
  eval_rollout_epoch: 1 # set eval_rollout_epoch > 0 when enable runner.only_eval or runner.val_check_interval > 0

  reward_type: chunk_level
  logprob_type: chunk_level
  entropy_type: token_level

  update_epoch: 4
  adv_type: gae
  loss_type: actor_critic
  loss_agg_func: "token-mean" 
  kl_beta: 0.0
  entropy_bonus: 0
  clip_ratio_high: 0.2
  clip_ratio_low: 0.2
  clip_ratio_c: 3.0
  value_clip: 0.2
  huber_delta: 10.0

  gamma: 0.99
  gae_lambda: 0.95

  filter_rewards: False
  rewards_lower_bound: -100
  rewards_upper_bound: 100
  # params for generation
  sampling_params:
    do_sample: True
    temperature_train: 1.0
    temperature_eval: 0.6
    top_k: 50
    top_p: 1.0
    repetition_penalty: 1.0
    add_BOS: False

  # length argument for autoregressive sampling
  # max length means max amount of tokens to generate
  length_params:
    max_new_token: null
    max_length: 1024
    min_length: 1

env:
  group_name: "EnvGroup"
  channel:
    name: "env_buffer_list"
    queue_name: "obs_buffer"
    queue_size: 0
  enable_offload: False

  train:
    total_num_envs: 16
    max_episode_steps: 480 # max episode steps for truncation
    max_steps_per_rollout_epoch: 480
    reward_coef: ${algorithm.reward_coef}
  eval:
    total_num_envs: 10
    auto_reset: True
    ignore_terminations: True
    max_episode_steps: 480
    max_steps_per_rollout_epoch: 480
    group_size: 1
    use_fixed_reset_state_ids: True
    use_ordered_reset_state_ids: True
    is_eval: True
    video_cfg:
      save_video: True
      video_base_dir: ${runner.logger.log_path}/video/eval

rollout:
  group_name: "RolloutGroup"
  channel:
    name: ${env.channel.name}
    queue_name: "action_buffer"
    queue_size: 0
  mode: "colocate"
  generation_backend: "huggingface"
  recompute_logprobs: False
  unnorm_key: libero_10
  enable_offload: False
  pipeline_stage_num: 1

  model:
    model_path: "/root/autodl-tmp/RLinf-Pi0-LIBERO-Long-SFT"
    precision: ${actor.model.precision}

actor:
  group_name: "ActorGroup"
  channel:
    name: ${env.channel.name}
    queue_name: "replay_buffer"
    queue_size: 0
  training_backend: "fsdp"
  micro_batch_size: 2
  global_batch_size: 16 # maximum : global batch size = micro batch size * actor num_processes.
  seed: 42
  enable_offload: False

  # Override the default values in model/pi0
  model:
    model_path: "/root/autodl-tmp/RLinf-Pi0-LIBERO-Long-SFT"
    add_value_head: True
    # openpi specific parameters
    openpi:
      detach_critic_input: True

  optim:
    lr: 5.0e-6
    value_lr: 1.0e-4
    adam_beta1: 0.9
    adam_beta2: 0.95
    adam_eps: 1.0e-05
    clip_grad: 2.0
    weight_decay: 0.01
    critic_warmup_steps: 0

  # Override the default values in training_backend/fsdp
  fsdp_config:
    strategy: "fsdp"
    sharding_strategy: "no_shard"
    gradient_checkpointing: False # for openpi, gradient checkpointing is not supported, please do not change this value
    mixed_precision:
      param_dtype: ${actor.model.precision}
      reduce_dtype: ${actor.model.precision}
      buffer_dtype: ${actor.model.precision}

reward:
  use_reward_model: False

critic:
  use_critic_model: False

```

### Log file

You can find the log file in `logs/` folder or the $output_dir/$experiment_name folder (defined in the yaml config) if you are using our example scripts.

Log file:

If you cannot find the log, please provide the full log messages here.
```
<paste log here>
```

### Environment

python -V:
uv pip list:
nvidia-smi:
RLinf version:
Docker image tag: 

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

**Comment by chenkang455** [comment]
Hi, @tobottyx, I noticed your micro_batch_size is set to 2. Since we use the loss mask to filter invalid state-action pairs, a batch size of 2 is highly likely to result in empty data if both samples are masked out, causing the NaN errors. There are two solutions:

1. **Set the micro-bs larger.** If you don't have much memory, you can refer this part: https://rlinf.readthedocs.io/en/latest/rst_source/examples/pi0.html#:~:text=%E2%AD%90%202.4%20Minimum%20Test%20Case%20%E2%AD%90

<img width="1544" height="878" alt="Image" src="https://github.com/user-attachments/assets/2996dd99-03be-49a7-a281-194fca9fb49a" />

2. You can consider **using the `ignore_terminations=True` mode**, which will not trigger the `loss-mask`:
```
train:
  total_num_envs: 64
  max_episode_steps: 480 
  max_steps_per_rollout_epoch: 480
  reward_coef: ${algorithm.reward_coef}
  ignore_terminations: True # ⭐add here ⭐
```

**Comment by tobottyx** [comment]
Thank you very much. I have redefined the reward function. Could this be related to the reward values being too small? Also, what effect does ignore_terminations: True have? I don't quite understand it.

**Comment by chenkang455** [comment]
@tobottyx No, the NaN issue is not related to your specific values. It mainly occurs because the amount of data in a single micro-batch is too small. Due to our masking mechanism, empty values can appear within a batch, leading to NaNs.

Setting ``ignore_terminations = False`` means that once a task succeeds, we discard the subsequent trajectories during the update process using the mask mechanism. Conversely, setting ``ignore_terminations = True`` means the system will continue running regardless of task success, and the loss-masking mechanism will not be activated. 

If you set the  ``ignore_terminations = True``, all loss-mask will be True and I think the nan problem can be alleviated.

<img width="1474" height="740" alt="Image" src="https://github.com/user-attachments/assets/f5de127f-a5e7-4c1a-ab6d-877867d3477b" />

**Comment by tobottyx** [comment]
I am encountering severe training instability and immediate policy collapse during the early stages of PPO training. Despite using a conservative learning rate (5e-6), the logs indicate that the policy updates are too aggressive. Specifically, train/actor/approx_kl consistently exceeds 0.09-0.12 (far above the typical 0.01-0.03 range), and train/actor/grad_norm reaches as high as 48.5, suggesting that gradient clipping might not be effectively constraining the updates or the reward scale is too large. Consequently, the agent's performance is degrading rapidly rather than improving (e.g., env/success_once dropped from 45% to 37% and env/return plummeted from 0.36 to 0.10 within just 8 global steps). Additionally, the training throughput is extremely low (~24 mins/step). Could this be caused by unnormalized rewards, excessive update_epochs, or a specific configuration issue with the gradient clipper?

**Comment by tobottyx** [comment]
defaults:
  - env/libero_10@env.train
  - env/libero_10@env.eval
  - model/pi0@actor.model
  - training_backend/fsdp@actor.fsdp_config
  - override hydra/job_logging: stdout

hydra:
  run:
    dir: .
  output_subdir: null
  searchpath:
    - file://${oc.env:EMBODIED_PATH}/config/

cluster:
  num_nodes: 1
  component_placement:
    actor: 0,
    env: 0,
    rollout: 1

runner:
  task_type: embodied
  logger:
    log_path: "../results"
    project_name: rlinf
    experiment_name: "test_openpi"
    logger_backends: ["tensorboard"] # wandb, swanlab

  max_epochs: 1000
  max_steps: -1

  only_eval: False
  eval_policy_path: null # Optional: .pt file or None, if None, will use the checkpoint in rollout.model.model_path
  val_check_interval: -1
  save_interval: 40
  seq_length: 2048
  max_prompt_length: 50

  resume_dir: null

algorithm:
  normalize_advantages: True
  kl_penalty: kl  # how to estimate kl divergence: kl or kl_penalty
  group_size: 1
  reward_coef: 1.0
  rollout_epoch: 4
  eval_rollout_epoch: 1 # set eval_rollout_epoch > 0 when enable runner.only_eval or runner.val_check_interval > 0

  reward_type: chunk_level
  logprob_type: chunk_level
  entropy_type: token_level

  update_epoch: 4
  adv_type: gae
  loss_type: actor_critic
  loss_agg_func: "token-mean" 
  kl_beta: 0.0
  entropy_bonus: 0
  clip_ratio_high: 0.2
  clip_ratio_low: 0.2
  clip_ratio_c: 3.0
  value_clip: 0.2
  huber_delta: 10.0

  gamma: 0.99
  gae_lambda: 0.95

  filter_rewards: False
  rewards_lower_bound: -100
  rewards_upper_bound: 100
  # params for generation
  sampling_params:
    do_sample: True
    temperature_train: 1.0
    temperature_eval: 0.6
    top_k: 50
    top_p: 1.0
    repetition_penalty: 1.0
    add_BOS: False

  # length argument for autoregressive sampling
  # max length means max amount of tokens to generate
  length_params:
    max_new_token: null
    max_length: 1024
    min_length: 1

env:
  group_name: "EnvGroup"
  channel:
    name: "env_buffer_list"
    queue_name: "obs_buffer"
    queue_size: 0
  enable_offload: False

  train:
    total_num_envs: 16
    max_episode_steps: 480 # max episode steps for truncation
    max_steps_per_rollout_epoch: 480
    reward_coef: ${algorithm.reward_coef}
  eval:
    total_num_envs: 10
    auto_reset: True
    ignore_terminations: True
    max_episode_steps: 480
    max_steps_per_rollout_epoch: 480
    group_size: 1
    use_fixed_reset_state_ids: True
    use_ordered_reset_state_ids: True
    is_eval: True
    video_cfg:
      save_video: True
      video_base_dir: ${runner.logger.log_path}/video/eval

rollout:
  group_name: "RolloutGroup"
  channel:
    name: ${env.channel.name}
    queue_name: "action_buffer"
    queue_size: 0
  mode: "colocate"
  generation_backend: "huggingface"
  recompute_logprobs: False
  unnorm_key: libero_10
  enable_offload: False
  pipeline_stage_num: 1

  model:
    model_path: "/root/autodl-tmp/RLinf-Pi0-LIBERO-Long-SFT"
    precision: ${actor.model.precision}

actor:
  group_name: "ActorGroup"
  channel:
    name: ${env.channel.name}
    queue_name: "replay_buffer"
    queue_size: 0
  training_backend: "fsdp"
  micro_batch_size: 64
  global_batch_size: 64 # maximum : global batch size = micro batch size * actor num_processes.
  seed: 42
  enable_offload: False

  # Override the default values in model/pi0
  model:
    model_path: "/root/autodl-tmp/RLinf-Pi0-LIBERO-Long-SFT"
    add_value_head: True
    # openpi specific parameters
    openpi:
      detach_critic_input: True

  optim:
    lr: 5.0e-6
    value_lr: 1.0e-4
    adam_beta1: 0.9
    adam_beta2: 0.95
    adam_eps: 1.0e-05
    clip_grad: 2.0
    weight_decay: 0.01
    critic_warmup_steps: 0

  # Override the default values in training_backend/fsdp
  fsdp_config:
    strategy: "fsdp"
    sharding_strategy: "no_shard"
    gradient_checkpointing: False # for openpi, gradient checkpointing is not supported, please do not change this value
    mixed_precision:
      param_dtype: ${actor.model.precision}
      reduce_dtype: ${actor.model.precision}
      buffer_dtype: ${actor.model.precision}

reward:
  use_reward_model: False

critic:
  use_critic_model: False

**Comment by chenkang455** [comment]
@tobottyx  I compared your config and our provided config:

<img width="1936" height="232" alt="Image" src="https://github.com/user-attachments/assets/4b0fba32-2667-4723-8395-e467e9415e99" />


<img width="2090" height="204" alt="Image" src="https://github.com/user-attachments/assets/c95e85a9-e7b6-4d6d-9d09-3274c0799975" />

<img width="2136" height="178" alt="Image" src="https://github.com/user-attachments/assets/32b97d9a-fa46-4265-b389-a66abb6cb2ac" />

Can you refer to this part for some guidance? 

<img width="1504" height="1420" alt="Image" src="https://github.com/user-attachments/assets/d591ee83-20d2-4a0e-bc98-e951c3d46a11" />

**Comment by BitCalSaul** [comment]
Hi, @chenkang455 

I have a question about the effect of max_episode_steps. Most example configs seem to use values around 256 or 512. Does increasing max_episode_steps mainly affect:

GPU memory usage (due to longer trajectories), and training time per iteration?

Is there any recommended minimum value? For example, would setting it to something smaller like 100 cause issues?
